mysql-test/suite/mysql_migration/mysql_migration.test

```diff
@@ -0,0 +1,206 @@
+# 2. mysql_migration.sql defines the databases/tables populated into the datadir
+#    Theoretically these data should reflect all the incompatibilities to be tested
+#
+let $POPULATE_MYSQL_DATADIR= suite/mysql_migration/populate_mysql_datadir.sh;
+let $MYSQL_COMMANDS= suite/mysql_migration/mysql_migration.sql;
+let $TEST_MYSQL_VER= 5.7;
+let $MYSQLD_DATADIR= `select @@datadir`; # For clean up
+
+# Regex note:
+# 1. For some reasons --replace_regex write \n as it is instead of new line. Therefore
+#    use a workaround to also capture the \n as a capturing group and put back into the
+#    replacement string
+# 2. When masking the container id, seems tat --regex_regex does not allow pattern matching
+#    preceding lines. Thus do a bind search on anything continues string longer than 40.
+--replace_regex /(?<=sha256:).*(\n)/<MASKED_CID>\1/  /^[a-zA-Z0-9]{40,}/<MASKED_CID>/
+--exec $POPULATE_MYSQL_DATADIR $TEST_MYSQL_VER $MYSQLTEST_VARDIR/tmp $MYSQL_COMMANDS 2>&1
+
+# Restart the server from the populated pseudo MySQL data directory
+let $restart_parameters=--datadir=$MYSQLTEST_VARDIR/tmp/mysql;
+--source include/restart_mysqld.inc 
+
+--echo #
+--echo # Run mariadb-upgrade
+--echo #
+
+# In fact some incompatibility errors may already appear during the upgrade run
+--exec $MYSQL_UPGRADE --force 2>&1
+
+# Restart the server and run mariadb-upgrade one more time to make sure that errors appeared
+# in the last run are resolved (e.g. column number mismatch for mysql sys.xxx views)
+--source include/restart_mysqld.inc 
+--exec $MYSQL_UPGRADE --force 2>&1
+
+--echo #
+--echo # Test that MariaDB can access tables partitioned with "native" InnoDb partition handler
+--echo #
+
+#
+# MariDB doc:
+#   MariaDB 10.3 does not support MySQL 5.7's “native” InnoDB partitioning handler.
+#
+# Please make sure the coresponding data exists in MySQL datadir via $POPULATE_MYSQL_DATADIR script
+#
+
+use idb_partition_db1;
+
+SHOW TABLES;
+
+--error ER_NO_SUCH_TABLE_IN_ENGINE # Allow the test to continue. Remove once this incompatibility is resolved
+select * from idb_partition_t1;
+
+# Confirm that MariaDB can still apply partitioning to the table as usual
+--error ER_NO_SUCH_TABLE_IN_ENGINE # Allow the test to continue. Remove once this incompatibility is resolved
+ALTER TABLE idb_partition_t1
+    PARTITION BY HASH(id)
+    PARTITIONS 8;
+
+--error ER_NO_SUCH_TABLE_IN_ENGINE # Allow the test to continue. Remove once this incompatibility is resolved
+select * from idb_partition_t1;
+
+--echo #
+--echo # Test that MariaDB can access tables compressed using MySQL punch-hole compression
+--echo #
+
+#
+# Please make sure the corresponding data exists in MySQL datadir via $POPULATE_MYSQL_DATADIR script
+#
+
+use idb_page_comp_db1;
+
+SHOW TABLES;
+
+--error ER_TABLE_CORRUPT # Allow the test to continue. Remove once this incompatibility is resolved
+select count(*) as EXPECT_3000 from idb_page_comp_t1;
```

Comment:
Do we really want to support this file format? The way the page is compressed probably differs from both already existing implementations of `page_compressed` in MariaDB. (The second file format was introduced in 6b6fa3cdb16ae7b4bc9e307c7d9b9012a055548c.)

Just like with anything that uses `CREATE TABLESPACE`, the solution here could be to require such tables to be rebuilt before upgrading.

Reply:
> Do we really want to support this file format?

> Do we really want to implement the MySQL 5.7 way of encrypting tables?

According to [the docs on upgrading from MySQL → MariaDB](https://mariadb.com/kb/en/upgrading-from-mysql-to-mariadb/):

“you can **in most cases** just uninstall MySQL [5.7] and install MariaDB and you are good to go. There is no need to dump and restore databases. As with any upgrade, we recommend making a backup of your data beforehand.”

The phrase “in most cases” is doing a lot of work there.

One of the purposes of this PR is to _clearly enumerate, document, and track_ the cases where this does not Just Work, but where data will be lost or corrupted in replacing MySQL 5.7 with MariaDB.

Reply:
Yes, it is good to document the incompatibilities. I would suggest to write a longer explanatory comment why this particular incompatibility is not likely to be removed.

The MySQL 5.7 encryption and sparse-file compression are similar to what was implemented and released in MariaDB 10.1 implemented earlier. I do not think it makes sense to support either format. In fact, I agree that [the practical usability of sparse-file compression is questionable](https://smalldatum.blogspot.com/2015/10/wanted-file-system-on-which-innodb.html). If you want that type of compression, I believe that the best option is thinly provisioned computational storage, such as a ScaleFlux device on which we performed some testing. It won hands down, both in terms of speed and compression ratio.

Supporting the MySQL 5.7 encryption could require implementing some plugin interface adapter. I would be reluctant to do that.

Those (as well as the `CREATE TABLESPACE`) are the only things that I think are technically infeasible to support. There are other compatibility issues, such as [MDEV-29253](https://jira.mariadb.org/browse/MDEV-29253) for supporting the `.frm` format that is used by MySQL 5.7 partitioning, which are easier and more feasible to fix.